### ISSUE TITLE
Make files reusable/selectable through File Fields and the rich text editor dialog

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#1307]
Besides being able to upload a file, users should be able to reuse existing images. A module like [FileField Sources](https://github.com/backdrop-contrib/filefield_sources) enables this functionality for fields, but it doesn't work (yet) with the dialog used by CKEditor for uploading files.

We should add the ability to reuse files directly to the File module. I think it would make the most sense if selecting an existing file opened a dedicated dialog for selecting an existing file. Out of the box, this dialog would probably be a page provided by Views that would have an exposed filter for searching files.

**Comment by klonos** [comment]
If memory serves well, there was this issue with reusing existing files where things went bananas when the file was renamed/removed. Can't remember respective d.org issue and no time have to :zzz:

**Comment by Graham-72** [comment]
I am very much looking forward to being able to use this feature. It is the one capability I really miss since the 'old days' of using CKEditor and IMCE in combination.

**Comment by docwilmot** [comment]
@quicksketch this has been sitting for a very long time and I suspect you're going to be the one to do it.

**Comment by daggerhart** [comment]
Here is a mockup I made to get this conversation going again.

![image-library-mockup](https://cloud.githubusercontent.com/assets/1205329/25966181/4bfddb40-3658-11e7-9377-b821b0365cf3.png)

The immediate problem I forsee with this mock is how demands demands much more vertical space. So there will end up being an interior scrolling window inside of that already-scrolling modal.

Thoughts/ feedback / alternative mockups welcome

**Comment by docwilmot** [comment]
My thoughts:
- Yay @daggerhart for pushing this
- the image list should be a view
- Options
  - The Image library tab could show the selected image from the library with a button/link to open a separate dialog (or replace contents of existing dialog) with the image list, OR 
  - The form above could be shifted to the side and append the View to the form

**Comment by olafgrabienski** [comment]
Nice progress! Regarding the needed space, it may be interesting to have a look at the Drupal File browser module, see this blog post with screencast: https://samuelmortenson.wordpress.com/2016/03/10/browsing-existing-files-and-images-in-drupal-8-with-file-browser/

**Comment by daggerhart** [comment]
This is a very minimal attempt at providing reusable files through the text editor dialog. Not to be confused with a real PR I expect get into core. More of a proof of concept.

I did the least amount of work possible to create an image library system that is usable in the rich test editor along side the "Image Upload" and "Image URL" features. It works, but has some problems and is a bit fragile in a few places.

The new field works by being almost exactly like the "Image URL" field. It just happens to have an embedded a view as the field's description (not ideal). There is some js that makes clicking on an image in the view populate this new field's value, and some other js that makes sure that value is saved.

Didn't do anything yet regarding CSS or styles. 

Thoughts?

**Comment by herbdool** [comment]
Thanks for the proof of concept! 

Indeed it's hard to think how it'll fit. I can see why in Drupal 7 the media browser is some tabs and a dialog. In this case I could imagine the Image library link could be a dialog with the library filling the dialog box with next/previous and select buttons. After selecting it brings the user back to the first view of the insert image dialog.

**Comment by olafgrabienski** [comment]
@daggerhart Had a quick look at your proof of concept. Great start! I uploaded some more images and noticed that we definitely need an exposed filter for the view.

**Comment by Al-Rozhkov** [comment]
Great to see progress here! Thank you @daggerhart!

My first thoughts that media browser have to be much wider, almost full width of viewport, like dialogs in views UI.

**Comment by daggerhart** [comment]
Thanks for the feedback. 

In order for me to clean this up into something usable, I would need to modify the modal form a good bit. Since we have jquery ui in core, I was thinking of using jquery.ui-tabs to handle the different tab-like options at the top:  "Image Upload", "Image URL", and "Image Library" would all be tabs.

**Comment by daggerhart** [comment]
Opinions please:

How do we feel about a media library based on a new node type? I've been working locally on this approach and need to get a feel for its acceptability before continuing.

My main reasoning is that w/ a node type we inherit a ton of free functionality from the core node system. I can see pros and cons, just interested in your thoughts.

**Comment by docwilmot** [comment]
Why not an Entity type? 

And, curious, what aspects of the media browser idea would benefit from having media items be nodes/entities?

And also, since youre reviving this (thanks again), maybe we need to create a list of wants/needs for a media browser experience up front to guide the development? Like @jenlampton did for forums:
[Must haves](https://github.com/backdrop-ops/forum.backdropcms.org/issues/4)
[Nice-to-haves](https://github.com/backdrop-ops/forum.backdropcms.org/issues/5)

**Comment by daggerhart** [comment]
Lists of Must/Nice-to haves would be great!  I can find time to work on this, but don't want to go too far in the wrong direction.

My thought process (open to criticism/suggestion):

I think a new entity type is ideal, but thinking it through I can't help but conclude that each type of media becomes its own bundle resulting in something very similar to the D7 media module + all extensions. That sounds overly complicated to me, but maybe there is another outcome?

Personally, I'd like a super-bundle. Something with an interface where I upload any file of any type, and the system automatically knows what type of media it is and can handle the output appropriately. But that's more of a UX thing, and is likely possible with the many-bundle entity type approach.

**Comment by docwilmot** [comment]
It would be great to have input from @jenlampton @Graham-72 re the list of needs/wants. And @quicksketch re the technical direction. Guys, chime in please? A simple built-in media browser would be another great WP-like feature to distinguish us from D8 even more, and @daggerhart is volunteering to lead it.

**Comment by Graham-72** [comment]
I think my main need is met by installing IMCE, though it may not work as elegantly as one might wish. It adds a 'file browser' link to the Insert Image popup and this enables me to select an image from the server. Is the image file already an entity, made one when first uploaded? Can it be re-used in another node? I don't know enough about these things.
![edit page - browse server](https://cloud.githubusercontent.com/assets/10559274/26164983/57a85d44-3b26-11e7-9bec-de3b2f4ef37d.png)
![file browser 1](https://cloud.githubusercontent.com/assets/10559274/26165001/66608244-3b26-11e7-9a4a-4d741ec1e545.png)

**Comment by docwilmot** [comment]
@Graham-72 Glad IMCE works for you. I tagged you here since you asked (repeatedly) for this feature in the past, not re the entity/node thing (I dont get it myself) but to get an idea of what features you would want/need your users to have if IMCE-like browser were to be added to core. A short list would be nice.

p.s.: In https://github.com/backdrop/backdrop-issues/issues/1307#issuecomment-300886430 above I suggested two options for launching the built-in browser: either a link to "browse" just like IMCE does OR a new tab like @daggerhart suggested in the image in https://github.com/backdrop/backdrop-issues/issues/1307#issuecomment-300881645. Which one would you prefer?

**Comment by Graham-72** [comment]
I think we need two links (tabs?) of equal status and very clearly labelled as to their function. One might be 'Upload a new image', the other 'Reuse an uploaded image'. Both would open the corresponding browser-type window, not necessarily of similar appearance because there may be merit in making it very clear that one is showing file images on the user's device, the other showing files on the server.

There also needs to be a field for direct entry of a URL for those expert enough to tackle this. Perhaps this also should provide a preview of the selected file?

I will try and think of any other feature that might be helpful, perhaps look and see the WP user experience!

**Comment by jenlampton** [comment]
Thanks everyone for reviving this issue!  It's an important feature :)

It sounds like what @Graham-72 is describing is the [filefield_sources module](https://www.drupal.org/project/filefield_sources). It sounds like what @daggerhart is describing is the [file_entity module](https://www.drupal.org/project/file_entity)  - some/most of which we already have in Backdrop core.

The one use-case that these two solutions don't already account for is the select-a-file from the server experience we get from IMCE. This has been avoided in Drupal because it's a security issue if improperly configured, and allowing people to manipulate files on disk bypasses all the records for those files in the database, leaving Drupal unaware of what files are supposed to be "managed". If we could create a similar experience - a file-on-disk-viewer - but then handle the management of those files in a drupal-y way, we might be able to prevent people from needing to use IMCE, and the problems associated with doing so.

I wonder if this use-case is a nice-to-have or a must-have. If we can provide a nice asset library experience, will the desire to manipulate files on disk disappear? @daggerhart in WordPress is the IMCE experience still important? @Graham-72 would you still want to select files from disk if you had a library you could use instead?

**Comment by Graham-72** [comment]
> would you still want to select files from disk if you had a library you could use instead?

I think a library would be better. I would rather my users did not see all the files, only those that might be useful and reusable. Do image files get added to the library when first used? It would be a shame if it were a two stage procedure (upload image to library, then separately insert it to text).

**Comment by docwilmot** [comment]
@jenlampton absolutely agreed with most youve said. I believe that simply uploading and linking to files is a no-go; this implementation would be to allow inserting links to files *which have already been uploaded*. Upload would continue to be through the existing `backdrop_image` plugin. 

We have already started down the road of asset management with our Files view; this would just be exposing this same list of files directly in CKEditor, and adding a click-to-insert-a-link-to-this-file functionality. I propose that initially this could just be images, but maybe later we can expand to videos, documents, whatever.

This is a essential feature of WP IMHO. And it doesnt exist in Drupal because its been debated to exhaustion (also my opinion). But the fact that IMCE is the **11th most downloaded module** on D.org speaks to the fact that everyone would want this in core, much less 80%.

**Comment by jenlampton** [comment]
I totally agree. The reason people use IMCE is because we have nothing else! I think if we give people something else (something better!) the desire for that approach will dissipate. :)

>  I believe that simply uploading and linking to files is a no-go

I think we need to support this use case too, but people should do it through the through the existing backdrop_image plugin. (I think we're agreeing, but I want to re-iterate what @Graham-72 said: We shouldn't make people go to create content -> add file, first)

I would also love to see the file browser (to reuse files) on file upload fields. Whatever approach we take it should be the same in both places.

**Comment by olafgrabienski** [comment]
Can't add much new thoughts but wanted at least say that I'd be very glad if Backdrop would get a simple Media browser in core!

Regarding the node type question of @daggerhart there may be some interesting aspects in the discussion about [page-less nodes](https://github.com/backdrop/backdrop-issues/issues/100).

Re IMCE-like solution: I wouldn't expect Backdrop's core solution allowing to browse files in all kind of directories. To reuse uploaded files is totally fine in my opinion.

And yes, as @jenlampton said, the Media browser should work with the text editor **and** with fields.

**Comment by Graham-72** [comment]
I have just tried the task of adding an image using Wordpress and I have to say that from a user viewpoint it seems excellent. I was sceptical about mixing images and audio files in the same procedure but in practice it works really well.

The only thing I am not entirely happy with is their 'Add media' button being separate from the toolbar.

**Comment by serundeputy** [comment]
I like the third tab approach. With the third tab being `Image Library`.  The image library tab having some version/variation of the view here: `/admin/content/files` should be fine those are file entities and we should be able to have the view check what file entities the user has access to.

I'd say for now MVC lets get a usable `Image Library` that we can apply to Rich Text fields and image fields and tackle other media in another itteration.

So I'm on board with the direction @daggerhart has proposed. 👍

**Comment by daggerhart** [comment]
For additional context, I annotated a lot of details about the WordPress media library here:  https://gist.github.com/daggerhart/64e6748c81e7c29ad3d0c7db27ce4a1a

Didn't include all of that in this post so everyone didn't have to scroll past it all the time.

**Comment by docwilmot** [comment]
Based on @daggerhart's post above and IMCE's list of features, created a list of Must vs Nice-to haves and inserted into first post. Please comment and/or move things around in the list.

**Comment by daggerhart** [comment]
Created a PR of my work in progress for a node-based solution.

Currently implemented:

- Manage Media admin view `admin/content/media`
- Upload new media from that page ^
- Single generic file upload field
- "Dynamic" field formatter that shows images using the image formatter, and other files with the generic file formatter
- No title field: Media nodes automatically receive their title from the filename


Not implemented: 

* Anything involving the wysiwyg. That would require I start editing other module's files, so I stopped there to ask for feedback.

### Main Idea

The idea here is that we can make a "super-bundle" with a few simple fields at-most, that acts aware of the file-type that is uploaded. This leaves core devs with the burden of implementing the "smartness" of the node type, ideally to the benefit of users.

### Screenshot of Manage Media admin view

![media-library](https://cloud.githubusercontent.com/assets/1205329/26177853/793b9694-3b29-11e7-83f9-e149826c8d71.png)
---

### Gif (hard-G) of uploading a new media item from the Manage Media admin page
![media-library](https://cloud.githubusercontent.com/assets/1205329/26177875/8e67e6e4-3b29-11e7-9391-ac21453fefb4.gif)

**Comment by jenlampton** [comment]
> Manage Media admin view admin/content/media

We already have a page like this in core, admin/content/files. Can we adapt that files listing page for this purpose?

**Comment by Graham-72** [comment]
@docwilmot a possible extra item for the list of features it the ability to rotate images, perhaps according to the EXIF orientation flag as used by iPads etc. Probably we need this to work automatically when uploading an image with this flag, but also it might be useful to be able to rotate any uploaded image through 90 or 180 degrees.

I have a user who has just hit this problem when uploading a photo from her iPad. It looks the right way round on her iPad but is not right once uploaded. It is very confusing for a non-technical person.

Drupal has projects https://www.drupal.org/project/exif_orientation and https://www.drupal.org/project/image_exif_autorotate.

**Comment by daggerhart** [comment]
@Graham-72 I completely agree. I've run into the same issue as a site developer, and in other cases as a user when my iphone was taking all pictures upside down.  We should auto-rotate during upload, and also offer a simple per-image rotate option to the user.

Added to the list of nice-to-haves

**Comment by jenlampton** [comment]
@daggerhart I would make some progress on media for Backdrop 1.8. How can I be most helpful to you in this area?

**Comment by daggerhart** [comment]
@jenlampton Here is a little update:

I've gone through multiple attempts to implement this one feature (reusable files in file fields) but have hit some snag or other in each case.  So I've taken a bit of a step back to better familiarize myself with what is done in D7 w/ the File Entity & Media modules, as well the approaches D8 is taking.

Seems to me that this issue needs to be broken down into a few bite-sized tasks (most of these taken almost directly from the D7 File Entity and Media modules):

1. Provide a new route for creating adding files to the file system. Something like: `file/new`. This can later be extended for file types, like: `file/new/audio`
1. Provide a route to edit an existing file by, like: `file/[fid]/edit`. This can be later extended to allow for a "Replace file" mechanism. 
1. Create a new "list of files" (Media Library) view whose sole purpose is finding files within the system, and allowing the user to select one. 
1. Create a new form that provides 2 main features (similar to the file field sources module):
    1. Upload a new file (`file/new` route)
    1. Select existing file (Media Library view)
1. Create a new file field widget as a parallel option the current one.  This is what the Media module does in D7 as a simple way around having to re-do the whole file field, and bonus: provides backwards compatibility as existing file field UX will not be modified.
    * This new widget would dialog-open up the new form we created in item 3 above.
    * Upon saving this dialog, the selected file ID would be returned to the calling element.
1. Create a new dynamic file field formatter that attempts to handle output of any file appropriately for its file/mime type.

Once we have those things in place, I think we'll be ready to start rocking on the media library functionality.

What I need to do next is break these items into individual issues where none exist, or identify existing issues that are asking for these improvements. Then start knocking them out one at a time.

Currently, I'm a bit behind because I've helping organize a conference. But I have some time coming up soon I can dedicate to these items.

Questions: 

* Do you see any problem with this list? Anything I should consider differently?
* Any existing issues you can help me identify which fit the description of items in my list?

**Comment by jenlampton** [comment]
This generally sounds like a good plan to me.

> Do you see any problem with this list? Anything I should consider differently?

Your #4 `Create a new form` probably doesn't need to be a whole form, maybe just a new form API element that perhaps includes the other form API elements? 

> Any existing issues you can help me identify which fit the description of items in my list?

I'd recommend starting with https://github.com/backdrop/backdrop-issues/issues/1448, this issue's parent. That is the META issue we are currently using for this task. As you create new issues we should also be adding them the top post on that list, so they will also be easy to find later.

**Comment by quicksketch** [comment]
> Do you see any problem with this list? Anything I should consider differently?

This list is great. If this individual issue is too broad, let's make separate issues for the sub-tasks, but add them to the already-existing list at #1448

> Any existing issues you can help me identify which fit the description of items in my list?

1 and possibly 2 are #2633.
3, 4, and 5 are the issue we are in now. If needed to split this up, let's use this issue for 5 and make new issues for 3 and 4. Then add them to the meta at #1448
I don't think we have an issue for 6.

**Comment by daggerhart** [comment]
Thanks y'all!

I've created new issues for those items which didn't have them, updated both the meta task #1448 and my last comment with references to the issue, and changed the name of this issue to be more specifically about creating a new field field widget.

Did what I could to provide some general information for each issue, mainly to describe what order in which they should be approached and why, but I'm sure there is room for discussion and improvement. I expect a lot more smaller things to come up around these items, such as new permissions needed, but hopefully they will be easy to work out as the tasks progress.

First up (as I see it), determining the core file entity bundles #2632

**Comment by jenlampton** [comment]
How much of this meta was covered by https://github.com/backdrop/backdrop-issues/issues/3297 and https://github.com/backdrop/backdrop-issues/issues/3134?

**Comment by klonos** [comment]
Since this seems to be a meta-issue, I've added the `[META]` indication in the title. Also switched this to a task instead of a feature request.
